/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

syntax = "proto3";

package resdb.raft;

message Transaction{
  int32 id = 1;
  bytes data = 2;
  bytes hash = 3;
  int32 proxy_id = 4;
  int32 proposer = 5;
  int64 uid = 6;
  int64 create_time = 7;
  int64 seq = 9;
}

message Entry {
  uint64 term = 1;
  bytes command = 2;
}

message AppendEntries{
  uint64 term = 1;
  int32 leaderId = 2;
  uint64 prevLogIndex = 3;
  uint64 prevLogTerm = 4;
  repeated Entry entries = 5;
  uint64 leaderCommitIndex = 6;
}

message AppendEntriesResponse {
  uint64 term = 1;
  bool success = 2;
  int32 id = 3;
  uint64 lastLogIndex = 5;
}

message RequestVote {
  uint64 term = 1;
  int32 candidateId = 2;
  uint64 lastLogIndex = 3;
  uint64 lastLogTerm = 4;
}

message RequestVoteResponse {
  uint64 term = 1;
  bool voteGranted = 2;
  int32 voterId = 3;
}

message DirectToLeader {
  uint64 term = 1;
  int32 leaderId = 2;
}

enum MessageType {
  None = 0;
  AppendEntriesMsg = 1;
  AppendEntriesResponseMsg = 2;
  RequestVoteMsg = 3;
  RequestVoteResponseMsg = 4;
  DirectToLeaderMsg = 5;
}

